### ISSUE TITLE
Environment setting problem

### ISSUE THREAD

**Issue by egghead2630** [nctu-eva-lab/3DV-2022#4]
Hello, TAs.

When I tried to set up my environment by

                conda create --name py3d --file requirements.txt

However, I suffered from the below errors:
               
![2022-05-16 17-15-17 的螢幕擷圖](https://user-images.githubusercontent.com/37588572/168560013-610c18c7-8613-413b-bb5a-154c78ce3a93.png)

It seems like I do not have some channels, or maybe something else,  so that anaconda cannot find these packages, thus cause the error.

Below is my channel list:

![2022-05-16 17-18-04 的螢幕擷圖](https://user-images.githubusercontent.com/37588572/168560532-63e72264-693b-438d-aea1-de018db7c6b4.png)

Conda version:
                  conda 4.12.0

I've googled for some docs trying to solve this issue, but I didn't find a way out.
 

I wanna know what else do I need to complete this environment setting.

Thank you!

**Comment by CharlesChiuGit** [comment]
Some fo packages are actucally install via `pip install`, you can make another `requirements.txt` just for pip install and use this command to install those pip packages:

```bash
pip install -r /path/to/requirements.txt
```
As, for `pytorch` related packages, such as ` torchvision`, `torchaudio`, etc.
don't install it from `pip install`, use this command instead:

```bash
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch
```

**Comment by egghead2630** [comment]
Thanks for reply!

So we need to

First Step: delete these two lines from requirement.txt:

                    torchaudio=0.11.0=py310_cu113
                    torchvision=0.12.0=py310_cu113
Second step: 
                    pip install -r /path/to/requirements.txt
Third step:
                    conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch

Are the above steps correct?

**Comment by CharlesChiuGit** [comment]
Yes, this should do. 
Fell free to ask more if it doesn't solve the problem.

**Comment by egghead2630** [comment]
Hello again

I have tried the above steps, but when I tried:

                 pip install -r requirement.txt

It shows the below error:

                      ERROR: Invalid requirement: '_libgcc_mutex=0.1=main' (from line 4 of requirements.txt)

                      Hint: = is not a valid operator. Did you mean == ?

After some searching, I found this article:

[](https://stackoverflow.com/questions/58369759/pip-install-packages-failing-invalid-requirement-no-matching-distribution)

This suggests that the requirement.txt  we now have is a yml file that only fits "conda create", but not fits "pip install" command.

And the problem now is that, maybe we either need:

                    1. The anaconda channels that TAs used to solve the PackageNotFoundError, so that we can directly use the "conda create" command to build the environment.
                    
                    2. Some other ways to solve this issue.

Above is the problem, thank you!

**Comment by CharlesChiuGit** [comment]
@egghead2630 I have replace requirements.txt to requirements.yml, see the commit: [#9151182](https://github.com/nctu-eva-lab/3DV-2022/commit/9151182f71efbbb9a580b2a9fb3d83fe78171220).

See the rewrote readme [section](https://github.com/nctu-eva-lab/3DV-2022#conda-env) too.

This should work flawlessly, but it will take some time. Just be patient.

**Comment by egghead2630** [comment]
Hello!

I've tried the revised command and requirement.yml.

Everything goes fine until installing xatlas.

xatlas failed to build wheel and shows below error:

![2022-05-25 15-43-08 的螢幕擷圖](https://user-images.githubusercontent.com/37588572/170209978-d616de55-2342-4a74-9c84-4cb87f981b8d.png)

![2022-05-25 15-43-45 的螢幕擷圖](https://user-images.githubusercontent.com/37588572/170210023-509d6f4a-4306-46b1-b516-00dfde29a9f5.png)


![2022-05-25 15-43-57 的螢幕擷圖](https://user-images.githubusercontent.com/37588572/170210061-f1974f38-d4f8-4b61-a8ef-a998325ac2fc.png)

I've also tried 

                   /anaconda3/envs/py3d/bin/pip install xatlas==0.0.6

Same errors show up.

Is there any way to solve this problem?  Thank you again!

**Comment by CharlesChiuGit** [comment]
Next time try to put the logs in the code block too.
Using images to show logs is not a preferred way.

**Comment by CharlesChiuGit** [comment]
I just installed `xatlas` on another machine and it worked.

```bash
❯ pip install xatlas
Collecting xatlas
  Downloading xatlas-0.0.6-cp36-cp36m-manylinux2010_x86_64.whl (207 kB)
     |████████████████████████████████| 207 kB 1.5 MB/s
Installing collected packages: xatlas
Successfully installed xatlas-0.0.6
```

**Comment by CharlesChiuGit** [comment]
Maybe you need this too:
```bash
sudo apt-get install python3-dev build-essential
```

**Comment by egghead2630** [comment]
Sorry for the images, I would notice that next time.

By the way, I try to directly 

                   pip install xatlas
 
in the py3d environment.

This time it's installed all good, so I wonder it a python 3.10 issue or somewhat.

Thank you for the advises

**Comment by CharlesChiuGit** [comment]
I don't think it's a python3.10 issue, since I just reproduced the conda env with the exact same `requirements.yml` which is py3.10.
I think you just got bad luck got the corrupted `.whl` file when downloading it due to internet issue.

I suggest you create your own `requirements.yml` from your conda env and see if that matches the `requirements.yml` I provided, in case anything goes wrong.

**Comment by egghead2630** [comment]
Hi, TA

I have modified the requirements.yml that delete only 
                        
                         xatlas==0.0.6

Then everything goes fine. 

However when I tried to install it manually by
                        
                        /anaconda3/envs/py3d/bin/pip install xatlas

or update the environment by
                         
                       conda env update --name py3d --file requirements.txt

These two ways failed, errors are just like above posts.

Then I also tried the above commands:

                        sudo apt-get install python3-dev build-essential
                        /anaconda3/envs/py3d/bin/pip install --upgrade pip setuptools wheel


It didn't work either.

However, in py3d env, when I use merely 
               
                 pip install xatlas

It's installed nicely without a problem, but this command install xatlas at environment "base" but not "py3d"

I have also tried to install pyproject.toml, not working.


Since Above is a cmake failed, I wonder if this a cmake failed or gcc problem, here I provided my version for you to check, maybe I could catch up with TA's version and try it again:
                      
               gcc: GNU 5.4.0
               cmake:  3.23.0-rc2

Above I have tried my best fixing this problem, but I am still stuck.

Could you provide some ideas fixing this? Thank you.

**Comment by CharlesChiuGit** [comment]
Nice work! You did devote a lot of effort in it. Gotta give you credit for debugging this. 👏

1. 
> However, in py3d env, when I use merely
> ```pip install xatlas```
> It's installed nicely without a problem, but this command install xatlas at environment "base" but not "py3d"

That's weird. If you're already in the `py3d` conda env and you `pip install XXX` in it, then it should be in the `py3d` env when you called `which XXX`. Did you modify something in your `.bashrc` file? like adding something to `PATH`?

2.
> I have also tried to install pyproject.toml, not working.

You don't need to install that. `.toml` is a file type, not a package.

3.
> ```bash
> gcc: GNU 5.4.0
> cmake:  3.23.0-rc2
> ```
> Could you provide some ideas fixing this? Thank you.

My `gcc`, `cmake`, `pip`:

```bash
$ gcc --version # system level
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
gcc (Ubuntu 8.4.0-3ubuntu2) 8.4.0 # on another server

$ cmake --version # system level
cmake version 3.16.3

$ pip --version # in conda env: py3d
pip 22.1 from /home/{user}/anaconda3/envs/py3d/lib/python3.10/site-packages/pip (python 3.10)
$ which pip # in conda env: py3d
/home/odie/anaconda3/envs/py3d/bin/pip
```

**Comment by CharlesChiuGit** [comment]
If you solve the problem, please share your result and close the issue, thanks.

**Comment by egghead2630** [comment]
Oh! sorry for this,

I got a little busy these two days.

I've tried another way to solve the problem, not working though.

I will commit a final sum up for this issue maybe on Monday, and close the issue simultaneously.

Since I've tried every way I could on this problem, I think I would install the environment on another computer or workstation.

Thank you for your patience on this weird problem.

Best.

**Comment by CharlesChiuGit** [comment]
Autcally I don't think it's necessary to install `xatlas`, as I recall this assignment only needs `yaml` but not `toml`. 
The reason why `xatlas` is in the env might because `hydra` can take both `yaml` and `toml`; therefore `xatlas` is installed by `hydra`. 
Since I only wrote hydra configs in `yaml`, `toml` releated packages might not needed.

Despite the packages is not needed, I bet you learned a lot when you tried to solve the issue. And the skills you learned will serve you well when you start your own research in the upcoming future.

Happy coding! :)

**Comment by egghead2630** [comment]
Oh that's a good news,

I forget to check the train.py and eval.py 

there are actually no xatlas , maybe now I can finally start this :)

I will still give a sum-up on Monday, thank you TA.

By the way, do we need a good GPU for this project?

**Comment by CharlesChiuGit** [comment]
Nice!

The voxel part might take a lot of vram, but I think you can cut the `batch_size` to make it trainable.

**Comment by egghead2630** [comment]
This commit is a sum-up for the environment problem, hope to be helpful.

After the revised requirement.yml coming out, I have met the below problem when creating environment.

                           xatlas building wheel failed error

Here I list every way I have tried to solve this problem.

1. use some alternative way to install xatlas:

                   /anaconda3/envs/py3d/bin/pip install xatlas
                   conda env update --name py3d --file requirements.txt

It failed reasonably.

2. Follow the advises, update the build tools:
                    
                    sudo apt-get install python3-dev build-essential
                    /anaconda3/envs/py3d/bin/pip install --upgrade pip setuptools wheel

Somehow it didn't work, not quite know why.

3. Since my base environment install xatlas smoothly, I export my base environment as .yml and manually install the dependencies in my py3d environment, note that this takes "a lot of" time that I don't suggest you to follow this.

This way didn't work either, thus I give up installing xatlas in py3d, fortunately this package is really not necessary.

The reason why I didn't try to update my gcc and cmake version is that, since I have successfully installed xatlas in my base environment, I guess it's not the problem.

Above is the sum-up,  I have to thank TA here for trying to solve this problem together with me, that's a great encouragement.
